solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py

```diff
@@ -214,77 +364,236 @@ def _scale_tensor(tensor, range_min, range_max, scale_min, scale_max):
+          sparse = tf.contrib.layers.sparse_column_with_hash_bucket(
+              name, 
+              hash_bucket_size=transform_config['hash_bucket_size'])
+          learn_feature = tf.contrib.layers.embedding_column(
+              sparse, 
+              dimension=train_config['vocab_stats'][name]['n_classes'])
+        elif transform_name == 'embedding':
+          sparse = tf.contrib.layers.sparse_column_with_integerized_feature(
+              name, 
+              bucket_size=train_config['vocab_stats'][name]['n_classes'])
+          learn_feature = tf.contrib.layers.embedding_column(
+              sparse, 
+              dimension=transform_config['embedding_dim'])
+        elif transform_name == 'one_hot' or transform_name == None:
+          sparse = tf.contrib.layers.sparse_column_with_integerized_feature(
+              name, 
+              bucket_size=train_config['vocab_stats'][name]['n_classes'])
+          learn_feature = tf.contrib.layers.one_hot_column(sparse)
+        else:
+          raise ValueError('For DNN modles, only hash_embedding, '
+                           'hash_one_hot, embedding, and one_hot transforms '
+                           'are supported.')
+      elif is_linear_model(args.model_type):
+        if transform_name == 'sparse' or transform_name == None:
+          learn_feature = tf.contrib.layers.sparse_column_with_integerized_feature(
+              name, 
+              bucket_size=train_config['vocab_stats'][name]['n_classes'])
+        elif transform_name == 'hash_sparse':
+          learn_feature = tf.contrib.layers.sparse_column_with_hash_bucket(
+              name, 
+              hash_bucket_size=transform_config['hash_bucket_size'])
+        else:
+          raise ValueError('For linear models, only sparse and '
+                           'hash_sparse are supported.')
+
+      #Save the feature
+      feature_columns.append(learn_feature)
+  return feature_columns
+
+
+
+# ==============================================================================
+# Building the TF learn estimators
+# ==============================================================================
 
-def produce_feature_engineering_fn(metadata, transform_config, schema_config):
-  """Makes a feature_engineering_fn for transforming the numerical types. 
 
-  This is called with the output of the 'input_fn' function, and the output of
-  this function is given to tf.learn to further process. This function extracts
-  the ids tensors from ml.features.FeatureMetadata.parse_features and throws
-  away the values tensor.
+def get_vocabulary(preprocess_output_dir, name):
+  vocab_file = os.path.join(preprocess_output_dir, CATEGORICAL_ANALYSIS % name)
+  if not ml.util._file.file_exists(vocab_file):
+    raise ValueError('File %s not found in %s' % (CATEGORICAL_ANALYSIS % name, preprocess_output_dir))
+
+  df = pd.read_csv(StringIO(ml.util._file.load_file(vocab_file)),
+                   header=0)
```

Comment:
header=0 means using the first line as header. It seems in local_preprocess, you have "df.to_csv(None, header=False)" so no header?